### ISSUE TITLE
Broken pipe error

### ISSUE THREAD

**Issue by palansher** [vimeo/psalm#7502]
Hello!

Thank you for your work!

Maybe you have idea why I see amphp errors almost every time I change code in my VS Code
My work env. is: ssh remote on Ubuntu 18.04 + php 8.0 + https://github.com/psalm/psalm-vscode-plugin

```
composer info | grep psalm
lctrs/psalm-psr-container-plugin      1.7.0              Let Psalm understand better psr11 containers
psalm/plugin-phpunit                  0.16.1             Psalm plugin for PHPUnit
vimeo/psalm                           4.18.1             A static analysis tool for finding errors in PHP applications
weirdan/doctrine-psalm-plugin         v2.1.0             Stubs to let Psalm understand Doctrine better

composer info | grep amp
amphp/amp                             v2.6.1             A non-blocking concurrency framework for PHP applications.
amphp/byte-stream                     v1.8.1             A stream abstraction to make working with non-blocking I/O simple.
```

So every day I have:

```
#4. unknown(0):fwrite
#5. /home/vladp/dev/sites/gt-vladp-dev1/vendor/amphp/byte-stream/lib/ResourceOutputStream.php(204):send
#6. /home/vladp/dev/sites/gt-vladp-dev1/vendor/amphp/byte-stream/lib/ResourceOutputStream.php(157):write
#7. /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/LanguageServer/ProtocolStreamWriter.php(33):write
#8. /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/LanguageServer/ClientHandler.php(107):notify
#9. /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/LanguageServer/LanguageClient.php(60):logMessage
#10. /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/LanguageServer/LanguageServer.php(502):clientStatus
#11. /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/LanguageServer/LanguageServer.php(439):shutdown
#12. /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/LanguageServer/LanguageServer.php(120):Psalm\Internal\LanguageServer\{closure}
#13. /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/LanguageServer/EmitterTrait.php(84):emit
#14. /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/LanguageServer/ProtocolStreamReader.php(146):emitClose
#15. /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/LanguageServer/ProtocolStreamReader.php(72):Psalm\Internal\LanguageServer\{closure}
#16. unknown(0):send
```

![image](https://user-images.githubusercontent.com/24850253/151474011-ce056e0a-8688-4475-a0d2-0a31e329bfff.png)

![image](https://user-images.githubusercontent.com/24850253/151474039-2e2f0861-180e-47de-9745-45e0b99b4a08.png)

```
[2022-01-28T04:28:21.606442+03:00] CommonErrors.ERROR: fatal:1 Exception (0) Could not get class storage for gt\settingsinterface  on line 46 in /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/Provider/ClassLikeStorageProvider.php
#0 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ClassLikeAnalyzer.php(109): Psalm\Internal\Provider\ClassLikeStorageProvider->get()
#1 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/InterfaceAnalyzer.php(26): Psalm\Internal\Analyzer\ClassLikeAnalyzer->__construct()
#2 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/NamespaceAnalyzer.php(117): Psalm\Internal\Analyzer\InterfaceAnalyzer->__construct()
#3 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/NamespaceAnalyzer.php(70): Psalm\Internal\Analyzer\NamespaceAnalyzer->collectAnalyzableClassLike()
#4 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FileAnalyzer.php(314): Psalm\Internal\Analyzer\NamespaceAnalyzer->collectAnalyzableInformation()
#5 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FileAnalyzer.php(193): Psalm\Internal\Analyzer\FileAnalyzer->populateCheckers()
#6 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(362): Psalm\Internal\Analyzer\FileAnalyzer->analyze()
#7 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(619): Psalm\Internal\Codebase\Analyzer->Psalm\Internal\Codebase\{closure}()
#8 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(291): Psalm\Internal\Codebase\Analyzer->doAnalysis()
#9 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/LanguageServer/LanguageServer.php(349): Psalm\Internal\Codebase\Analyzer->analyzeFiles()
#10 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/LanguageServer/LanguageServer.php(188): Psalm\Internal\LanguageServer\LanguageServer->doAnalysis()
#11 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/LanguageServer/EmitterTrait.php(84): Psalm\Internal\LanguageServer\LanguageServer->Psalm\Internal\LanguageServer\{closure}()
#12 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/LanguageServer/ProtocolStreamReader.php(68): Psalm\Internal\LanguageServer\ProtocolStreamReader->emit()
#13 [internal function]: Psalm\Internal\LanguageServer\ProtocolStreamReader->Psalm\Internal\LanguageServer\{closure}()
#14 /home/vladp/dev/sites/gt-vladp-dev1/vendor/amphp/amp/lib/Coroutine.php(118): Generator->send()
#15 /home/vladp/dev/sites/gt-vladp-dev1/vendor/amphp/amp/lib/Internal/Placeholder.php(149): Amp\Coroutine->Amp\{closure}()
#16 /home/vladp/dev/sites/gt-vladp-dev1/vendor/amphp/amp/lib/Deferred.php(53): Amp\Promise@anonymous->resolve()
#17 /home/vladp/dev/sites/gt-vladp-dev1/vendor/amphp/byte-stream/lib/ResourceInputStream.php(101): Amp\Deferred->resolve()
#18 /home/vladp/dev/sites/gt-vladp-dev1/vendor/amphp/amp/lib/Loop/NativeDriver.php(327): Amp\ByteStream\ResourceInputStream::Amp\ByteStream\{closure}()
#19 /home/vladp/dev/sites/gt-vladp-dev1/vendor/amphp/amp/lib/Loop/NativeDriver.php(127): Amp\Loop\NativeDriver->selectStreams()
#20 /home/vladp/dev/sites/gt-vladp-dev1/vendor/amphp/amp/lib/Loop/Driver.php(138): Amp\Loop\NativeDriver->dispatch()
#21 /home/vladp/dev/sites/gt-vladp-dev1/vendor/amphp/amp/lib/Loop/Driver.php(72): Amp\Loop\Driver->tick()
#22 /home/vladp/dev/sites/gt-vladp-dev1/vendor/amphp/amp/lib/Loop.php(95): Amp\Loop\Driver->run()
#23 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php(480): Amp\Loop::run()
#24 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/Cli/LanguageServer.php(321): Psalm\Internal\Analyzer\ProjectAnalyzer->server()
#25 /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/psalm-language-server(7): Psalm\Internal\Cli\LanguageServer::run()
#26 {main}  

```

if I turn on TCP mode I have:

Code: -32603
```
[Info - 4:47:23 AM] Connection to server got closed. Server will restart.
[Error - 4:47:42 AM] Request textDocument/codeAction failed.
Message: InvalidArgumentException: Could not get file storage for /home/vladp/dev/sites/gt-vladp-dev1/src/paymenttodriver/settings/settingsinterface.php in /home/vladp/dev/sites/gt-vladp-dev1/vendor/vimeo/psalm/src/Psalm/Internal/Provider/FileStorageProvider.php:47
```
But I am sure I have enough space:
![image](https://user-images.githubusercontent.com/24850253/151474449-7c170a6e-bbfc-458c-8a14-58ef8fae0901.png)

full trace:
[errors.log](https://github.com/vimeo/psalm/files/7955173/errors.log)

Maybe I can get some advice from you!
thank you!

**Comment by palansher** [comment]
> Hey @palansher, can you reproduce the issue on https://psalm.dev ?

No Those are system errors :(

**Comment by weirdan** [comment]
Your error handler is misbehaving and reporting suppressed notices that were meant to be suppressed (see 
https://github.com/amphp/byte-stream/blob/10c420782bd805c4af634bab69b5dfdc14ab721c/lib/ResourceOutputStream.php#L78-L84)

These stack frames do not come from Psalm's code:
```
#1. /home/vladp/dev/sites/gt-vladp-dev1/src/Error/Handler/LegacyErrorHandler.php(104):getTraceAsString
#2. /home/vladp/dev/sites/gt-vladp-dev1/src/Error/Handler/LegacyErrorHandler.php(202):logErrors
#3. /home/vladp/dev/sites/gt-vladp-dev1/src/Error/Handler/LegacyErrorHandler.php(248):Gt\Error\Handler\{closure}
```

So you either need to disable the error handler or fix it to take `error_reporting()` value into account, similar to how Psalm does it:
https://github.com/vimeo/psalm/blob/1a6f968e157cadfe073e4992ff7dd062d67286ad/src/Psalm/Internal/ErrorHandler.php#L68

**Comment by palansher** [comment]
thanks for advice

> $written = @\fwrite($stream, $data, $chunkSize);

maybe the reason is that since php8 "@" works differently:
[https://www.php.net/manual/en/language.operators.errorcontrol.php](https://www.php.net/manual/en/language.operators.errorcontrol.php)

> Prior to PHP 8.0.0, the value of the severity passed to the custom error handler was always 0 if the diagnostic was suppressed. This is no longer the case as of PHP 8.0.0.


Anyway, in my dev environment, I must see all errors/warnings. So, probably I need to suppress warnings only to this particular file source in my error handler...


PS:

I am a PHP dummy but why just not check for NULL before \fwrite execution to avoid warning at all  :confused:  ....

> ```
>  // Error reporting suppressed since fwrite() emits E_WARNING if the pipe is broken or the buffer is full.
>                     // Use conditional, because PHP doesn't like getting null passed
>                     if ($chunkSize) {
>                         $written = @\fwrite($stream, $data, $chunkSize);
>                     } else {
>                         $written = @\fwrite($stream, $data);
>                     }
> ```

**Comment by palansher** [comment]
Solved in 4.19.0 release
